src/ixdat/readers/biologic.py

```diff
@@ -1,9 +1,212 @@
+"""This module implements the Reader for .mpt files made by BioLogic's EC-Lab software"""
+
+import re
+import time
+import numpy as np
+
 from . import TECHNIQUE_CLASSES
+from ..data_series import TimeSeries, ValueSeries
+from ..exceptions import ReadError
 
-ECMeasruement = TECHNIQUE_CLASSES["EC"]
+ECMeasurement = TECHNIQUE_CLASSES["EC"]
+delim = "\t"
+t_str = "time/s"
+timestamp_string_form = "%m/%d/%Y %H:%M:%S"  # like 07/29/2020 10:31:03
+regular_expressions = {
+    "N_header_lines": "Nb header lines : (.+)\n",
+    "timestamp_string": "Acquisition started on : (.+)\n",
+    "loop": "Loop ([0-9]+) from point number ([0-9]+) to ([0-9]+)",
+}
 
 
 class BiologicMPTReader:
-    def read(self):
-        # return ECMeasurement(**obj_as_dict)
-        pass
+    """A class to read .mpt files written by Biologic's EC-Lab.
+
+    read() is the important method - it takes the path to the mpt file as argument
+    and returns an ECMeasurement object (ec_measurement) representing that file.
+    The ECMeasurement contains a reference to the BiologicMPTReader object, as
+    ec_measurement.reader. This makes available all the following stuff, likely
+    useful for debugging.
+
+    Attributes:
+        path_to_file (Path): the location and name of the file read by the reader
+        n_line (int): the number of the last line read by the reader
+        place_in_file (str): The last location in the file read by the reader. This
+            is used internally to tell the reader how to parse each line. Options are:
+            "header", "column names", and "data".
+        header_lines (list of str): a list of the header lines of the files. This
+            includes the column name line. The header can be nicely viewed with the
+            print_header() function.
+        timestamp_string (str): The string identified to represent the t=0 time of the
+            measurement recorded in the file.
+        tstamp (str): The unix time corresponding to t=0, parsed from timestamp_string
+        ec_technique (str): The name of the electrochemical sub-technique, i.e.
+            "Cyclic Voltammatry Advanced", etc.
+        N_header_lines (int): The number of lines in the header of the file
+        column_names (list of str): The names of the data columns in the file
+        column_data (dict of str: np.array): The data in the file as a dict.
+            Note that the np arrays are the same ones as in the measurement's DataSeries,
+            so this does not waste memory.
+    """
+
+    def __init__(self):
+        """Initialize a Reader for .mpt files. See class docstring."""
+        self.path_to_file = None
+        self.n_line = 0
+        self.place_in_file = "header"
+        self.header_lines = []
+        self.timestamp_string = None
+        self.tstamp = None
+        self.ec_technique = None
+        self.N_header_lines = None
+        self.column_names = []
+        self.column_data = {}
+
+    def read(self, path_to_file):
```

Comment:
It seems mostly like a reader, such as this, could (maybe should) be a function and the object here serves only the purpose of maintaining parser state. And in reality, the main thing needed to turn it into a function, would be that the data containers (and possibly parser state) would need to be passed along to the sub-parser functions as arguments. So I guess that the real review comment is for you to consider for a second whether a function based implementation might in fact be "cleaner" and more testable. What you loose, I guess ,is the ability for the resulting object to refer back to the parser and view e.g. the filesname and the raw header lines and such. And this may be important enough to keep it like it is.

In any case, if you do keep it I would suggest to make the parser object "single-use", by having a `used` instance variable (or something with a better name) which is set at the end of `read` and checked at the start of `read`, to prevent a user accidentally using it more that once, which is sure to produce weird results.

Reply:
I think the Reader class setup is really nice. In addition to your reasons, I'll add that this was the easiest and most pleasant writing of a file parser I have ever experienced.

Ive added the `used` attribute, refactored to `file_has_been_read`, to avoid reading the file twice.

Reply:
Code representing this change is in the start of `read()`:
```
        if self.file_has_been_read:
            print(f"This {self.__class__.__name__} has already read {self.path_to_file}."
                  " Returning the measurement resulting from the original read. "
                  "Use a new Reader if you want to read another file.")
            return self.measurement
```
The behavior of attributes `file_has_been_read` and `measurement` are both documented in the class docstring.